File: api/internal/responses/responses.go
// Package responses pre-builds the small set of HTTP responses the server can
// return. The fraud-score endpoint has only six possible bodies — one per
// fraud count in {0..5} — so we format them once at startup and emit the
// pre-built byte slice on each request.
//
// This is the same trick the reference C solution uses: zero JSON marshaling
// in the hot path, just a memcpy.
package responses

import (
	"fmt"
)

// FraudScore[i] is the full HTTP/1.1 200 response (status line + headers + body)
// for `i` fraud votes among the top-5. The decision rule (matches the reference):
//
//	score    = i * 0.2
//	approved = i < 3
//
// Mapping:
//
//	0 → score 0.0, approved
//	1 → score 0.2, approved
//	2 → score 0.4, approved
//	3 → score 0.6, denied
//	4 → score 0.8, denied
//	5 → score 1.0, denied
//
// Threshold (frauds < 3 ↔ score < 0.6) matches DETECTION_RULES.md exactly.
var FraudScore [6][]byte

// Ready is the response for GET /ready.
var Ready []byte

// Status responses for error paths. NotFound, BadRequest, etc.
var (
	NotFound    []byte
	BadRequest  []byte
	TooLarge    []byte
	ServerError []byte
)

// Init must be called once at process startup before the server begins
// serving traffic.
func Init() {
	for i := 0; i <= 5; i++ {
		score := float32(i) * 0.2
		approved := "true"
		if i >= 3 {
			approved = "false"
		}
		body := fmt.Sprintf(`{"approved":%s,"fraud_score":%.4f}`, approved, score)
		FraudScore[i] = build200(body)
	}
	Ready = []byte("HTTP/1.1 200 OK\r\nContent-Length: 0\r\nConnection: keep-alive\r\n\r\n")
	NotFound = []byte("HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\nConnection: close\r\n\r\n")
	BadRequest = []byte("HTTP/1.1 400 Bad Request\r\nContent-Type: application/json\r\nContent-Length: 27\r\nConnection: close\r\n\r\n{\"error\":\"invalid_payload\"}")
	TooLarge = []byte("HTTP/1.1 413 Payload Too Large\r\nContent-Length: 0\r\nConnection: close\r\n\r\n")
	ServerError = []byte("HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\nConnection: close\r\n\r\n")
}

func build200(body string) []byte {
	return []byte(fmt.Sprintf(
		"HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: %d\r\nConnection: keep-alive\r\n\r\n%s",
		len(body), body,
	))
}

File: api-rust/src/responses.rs
// Pre-built HTTP/1.1 responses. Six fraud-score bodies (one per top-5 fraud
// vote count 0..5) and a small set of error responses. Built once at startup,
// emitted as raw byte slices on the hot path. Body shape is byte-identical to
// api/internal/responses/responses.go.

use std::sync::OnceLock;

static FRAUD_SCORE: OnceLock<[Vec<u8>; 6]> = OnceLock::new();
static READY: OnceLock<Vec<u8>> = OnceLock::new();
static NOT_FOUND: OnceLock<Vec<u8>> = OnceLock::new();
static BAD_REQUEST: OnceLock<Vec<u8>> = OnceLock::new();
static TOO_LARGE: OnceLock<Vec<u8>> = OnceLock::new();
static SERVER_ERROR: OnceLock<Vec<u8>> = OnceLock::new();

pub fn init() {
    let _ = FRAUD_SCORE.set([
        build200(0),
        build200(1),
        build200(2),
        build200(3),
        build200(4),
        build200(5),
    ]);
    let _ = READY.set(
        b"HTTP/1.1 200 OK\r\nContent-Length: 0\r\nConnection: keep-alive\r\n\r\n".to_vec(),
    );
    let _ = NOT_FOUND.set(
        b"HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\nConnection: close\r\n\r\n".to_vec(),
    );
    let _ = BAD_REQUEST.set(
        b"HTTP/1.1 400 Bad Request\r\nContent-Type: application/json\r\nContent-Length: 27\r\nConnection: close\r\n\r\n{\"error\":\"invalid_payload\"}".to_vec(),
    );
    let _ = TOO_LARGE.set(
        b"HTTP/1.1 413 Payload Too Large\r\nContent-Length: 0\r\nConnection: close\r\n\r\n".to_vec(),
    );
    let _ = SERVER_ERROR.set(
        b"HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\nConnection: close\r\n\r\n"
            .to_vec(),
    );
}

#[inline]
pub fn fraud_score(frauds: usize) -> &'static [u8] {
    FRAUD_SCORE.get().expect("responses::init() not called")[frauds].as_slice()
}

#[inline]
pub fn ready() -> &'static [u8] {
    READY.get().expect("responses::init() not called").as_slice()
}

#[inline]
pub fn not_found() -> &'static [u8] {
    NOT_FOUND.get().expect("responses::init() not called").as_slice()
}

#[inline]
pub fn bad_request() -> &'static [u8] {
    BAD_REQUEST.get().expect("responses::init() not called").as_slice()
}

#[inline]
pub fn too_large() -> &'static [u8] {
    TOO_LARGE.get().expect("responses::init() not called").as_slice()
}

#[inline]
pub fn server_error() -> &'static [u8] {
    SERVER_ERROR.get().expect("responses::init() not called").as_slice()
}

fn build200(frauds: usize) -> Vec<u8> {
    let score = frauds as f32 * 0.2;
    let approved = if frauds < 3 { "true" } else { "false" };
    // Match Go's `%.4f` formatting.
    let body = format!(r#"{{"approved":{approved},"fraud_score":{score:.4}}}"#);
    let mut out = Vec::with_capacity(128 + body.len());
    out.extend_from_slice(b"HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: ");
    out.extend_from_slice(body.len().to_string().as_bytes());
    out.extend_from_slice(b"\r\nConnection: keep-alive\r\n\r\n");
    out.extend_from_slice(body.as_bytes());
    out
}
